### ISSUE TITLE
Is it possible to set an enum from an external source or load on demand (without dependsOn)?

### ISSUE THREAD

**Issue by benj** [brutusin/json-forms#42]
I want to set a select object to have a set of Text/Values as per an HTML class. The saved value should be the value of the select and the display just the Name e.g.

```
<select> 
<option value="file1.json">Peter</option>
<option value="file2.json">Mary</option>
</select>
```

Enums don't seem to offer this and I was thinking about creating an on the fly schema, but it wouldn't give the array structure as far as I can tell.

Idea is that when the data is saved you get

```
{
    "FileName":"file1.json"
}
```

rather than 
```
{
    "FileName":"Peter"
}
```
or 
```
{
    "Name":"Peter",
    "FileName":"file1.json"
}
```

**Comment by idelvall** [comment]
Sorry, but I can't understand the question.

**Comment by benj** [comment]
Sorry - let me rephrase. I want create the enum in the stub below from an external file at runtime.

```
{
  "$schema":"http://json-schema.org/draft-03/schema#",
  "type":"object",
  "properties":{
	"Shortlist":{
		"type":"string",
		"title":"Name",
		"description":"Handy descriptor to retrieve shortlist",
		"required":true
	},
    "Candidates":{
      "title":"Candidate",
      "type":"array",
      "items":{
        "type":"object",
        "properties":{
          "CandidateFile":{
            "type":"string",
            "title":"Candidate",
            "description":"Candidate",
            "required":true,
            "enum": [
              "Person1",
              "Person2",
              "Person3"
              ]
          }
        }
      }
    }
  }
}
```

e.g.
```
            "enum": "/url/to/load/data"
```

However ideally the enum I want to achieve is a key/value pair so that when loaded it is
```
            "enum": [
              "Person1":"FileName1.json",
              "Person2":"FileName2.json",
              "Person3":"FileName3.json"
              ]
```
so that when I getData() the return is 
```
{
    "Shortlist": "Name",
    "Candidates": [
        {
            "CandidateFile": "FileName1.json"
        },
        {
            "CandidateFile": "FileName2.json"
        }
    ]
}
```

**Comment by idelvall** [comment]
Sorry but it's still not clear:
1.  You have and external file defining what? Give me an example of that file please.
2.  What exactly do you want to achieve? My impression is that you want to let the user create some JavaScript objects using the information in this file, through the rendered html form, but again, i am not sure.

**Comment by benj** [comment]
1) The external file defines the enum as a key/value pair as follows 
```
{[
             { "FileName1.json":"Person1"},
              {"FileName2.json":"Person2"},
             { "FileName3.json":"Person3"}
]}
```

2) When the schema renders I want the select box that's created (format: string with enum from file) to display the value from the enum file, however when it exports data with getData() returned the key.

i.e. I select Person2 on the form, but the getData() output has FileName2.json.

Does that make more sense?

If I'm using pure HTML I can create a SELECT like this 
```
<select> 
<option value="FileName1.json">Person1</option>
<option value="FileName2.json">Person2</option>
</select>
```
So when I see it on the screen I see **Person1** but when I get the value I get **FileName1.json**

**Comment by idelvall** [comment]
ok i understand now. 

Supposing that you already know how to read the external information and you have an object instantiated with it, it's easy: 

The library can not do that directly but, you can decorate the data object returned by the form with your external information to construct a new object with the desired structure.

**Comment by benj** [comment]
OK -- yeah, know how to pull the external info in and create an object.
How do I apply the decorator?

**Comment by idelvall** [comment]
```javascript
<script>
// external
var externalEnum = {
  "FileName1.json":"Person1",
  "FileName2.json":"Person2",
  "FileName3.json":"Person3"
  }
  
// returned by form  
var data = {
    "Shortlist": "Some Entered Name",
    "Candidates": [
        {
            "CandidateFile": "Person1"
        },
        {
            "CandidateFile": "Person2"
        }
    ]
}  

// decoration
function decorate(formData, map){
	for (var i = 0; i < formData.Candidates.length; i++) {
		formData.Candidates[i].CandidateFile=map[formData.Candidates[i].CandidateFile];
	}
}

function invert(map){
	var ret = new Object();
	for (var entry in map) {
		var value = map[entry];
		ret[value] = entry;	
	}
	return ret;
}

decorate(data, invert(externalEnum));

alert(JSON.stringify(data))  
  </script>
````

**Comment by benj** [comment]
Thanks so much. I'll give it a go.

**Comment by idelvall** [comment]
you're welcome mate!
